File: mobile/flutter_app/lib/features/exercises/domain/models/exercise_log.dart
class ExerciseLog {
  final int? id;
  final int? userId;
  final String dayIsoDate;
  final String exerciseName;
  final String? blockTitle;
  final int setNumber;
  final int? reps;
  final double? weightKg;
  final int? durationSeconds;
  final String? notes;
  final DateTime? loggedAt;

  const ExerciseLog({
    this.id,
    this.userId,
    required this.dayIsoDate,
    required this.exerciseName,
    this.blockTitle,
    required this.setNumber,
    this.reps,
    this.weightKg,
    this.durationSeconds,
    this.notes,
    this.loggedAt,
  });

  factory ExerciseLog.fromJson(Map<String, dynamic> json) {
    return ExerciseLog(
      id: json['id'] as int?,
      userId: json['user_id'] as int?,
      dayIsoDate: json['day_iso_date'] as String? ?? '',
      exerciseName: json['exercise_name'] as String? ?? '',
      blockTitle: json['block_title'] as String?,
      setNumber: json['set_number'] as int? ?? 1,
      reps: json['reps'] as int?,
      weightKg: (json['weight_kg'] as num?)?.toDouble(),
      durationSeconds: json['duration_seconds'] as int?,
      notes: json['notes'] as String?,
      loggedAt: json['logged_at'] != null
          ? DateTime.tryParse(json['logged_at'] as String)
          : null,
    );
  }

  Map<String, dynamic> toJson() => {
        'day_iso_date': dayIsoDate,
        'exercise_name': exerciseName,
        if (blockTitle != null) 'block_title': blockTitle,
        'set_number': setNumber,
        if (reps != null) 'reps': reps,
        if (weightKg != null) 'weight_kg': weightKg,
        if (durationSeconds != null) 'duration_seconds': durationSeconds,
        if (notes != null) 'notes': notes,
      };
}

File: backend/api/app/models/exercise_log.py
from datetime import datetime
from typing import Optional

from sqlalchemy import DateTime, Float, ForeignKey, Index, Integer, String, Text, func
from sqlalchemy.orm import Mapped, mapped_column

from app.db.base import Base


class ExerciseLog(Base):
    __tablename__ = "exercise_logs"
    __table_args__ = (
        Index("ix_exercise_logs_user_day", "user_id", "day_iso_date"),
    )

    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    user_id: Mapped[int] = mapped_column(
        Integer,
        ForeignKey("users.id", ondelete="CASCADE"),
        nullable=False,
    )
    plan_id: Mapped[Optional[int]] = mapped_column(
        Integer,
        ForeignKey("initial_plans.id", ondelete="SET NULL"),
        nullable=True,
    )
    day_iso_date: Mapped[str] = mapped_column(String(10), nullable=False)  # "2026-03-13"
    exercise_name: Mapped[str] = mapped_column(String(240), nullable=False)
    exercise_catalog_id: Mapped[Optional[int]] = mapped_column(
        Integer,
        ForeignKey("exercise_catalog.id", ondelete="SET NULL"),
        nullable=True,
    )
    block_title: Mapped[Optional[str]] = mapped_column(String(120), nullable=True)
    set_number: Mapped[int] = mapped_column(Integer, default=1, nullable=False)
    reps: Mapped[Optional[int]] = mapped_column(Integer, nullable=True)
    weight_kg: Mapped[Optional[float]] = mapped_column(Float, nullable=True)
    duration_seconds: Mapped[Optional[int]] = mapped_column(Integer, nullable=True)
    notes: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    logged_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        server_default=func.now(),
    )
